### ISSUE TITLE
[BUG]: stdout not collected from a crashing execution

### ISSUE THREAD

**Issue by OfekShilon** [compiler-explorer/compiler-explorer#7239]
### Describe the bug

Reported in [a linkedin discussion](https://www.linkedin.com/feed/update/urn:li:activity:7277040608915959809?commentUrn=urn%3Ali%3Acomment%3A%28activity%3A7277040608915959809%2C7277052131554082817%29&dashCommentUrn=urn%3Ali%3Afsd_comment%3A%287277052131554082817%2Curn%3Ali%3Aactivity%3A7277040608915959809%29): 
When executing this in CE -
```
#include <iostream>

int main() {
    std::cerr << "aaa\n";
    std::cout << "bbbb\n"; 

    int* ptr = nullptr;
    int x = *ptr;
}
```
Shows just `aaa`. 
This seems to be due to buffering, as if we add to the cout line -
```
    std::cout << "bbbb\n" << std::flush;
```
'bbbb' is shown as well.

Hopefully something can be done about it in node..

### Steps to reproduce

-

### Expected behavior

-

### Reproduction link

https://compiler-explorer.com/z/of49MKWz5

### Screenshots

Not applicable

**Comment by mattgodbolt** [comment]
This isn't a node thing: this is what the C runtime does unless overridden if it thinks the output is _not_ a TTY. We will have to somehow convince it its output is not to be buffered; I'm not sure how to do that. Definitely worth a look though! (complicated by it running in a sandbox which might make any tricks to force it to allocate a TTY more tricky?) thanks for filing this!

**Comment by OfekShilon** [comment]
Right. 
I'm going on vacation for 1.5w, will be happy to re-join the investigation when I'm back (but of course, don't wait up..)

**Comment by pavlix** [comment]
> This isn't a node thing: this is what the C runtime does unless overridden if it thinks the output is _not_ a TTY.

I didn't know it and didn't even believe you until I tried. So it finally is about buffering and therefore we can actually miss the output. This is an interesting inconsistency to me.

https://marc.info/?l=glibc-bug&m=98313957306297&w=4

Thank you for giving me a new perspective.

As for the possible solutions, one could probably use a pseudo tty. So the question won't be how to fix it but whether it's desirable in the context of compiler explorer, whether it should be configurable and what should be the default.

**Comment by OfekShilon** [comment]
Also, I can think of a few in-source workarounds in the meantime.  The easiest would be to add -
```
std::cout.setf(std::ios::unitbuf);
```
(or `setvbuf(stdout, NULL, _IONBF, 0)` for C) at the beginning of your `main`. 

You could also wrap the `main` body in try-catch, and in the catch clause call `std::flush` or similar.

**Comment by partouf** [comment]
Truth be told, I don't think we should do anything about this. Hiding normal behavior is not something anyone learns from.

**Comment by OfekShilon** [comment]
> Truth be told, I don't think we should do anything about this. Hiding normal behavior is not something anyone learns from.

When building and running on a terminal buffered IO *does* show for crashing apps.  This is perceived as 'normal' behavior, and if we could replicate that I think we should.

**Comment by OfekShilon** [comment]
TIL about stdbuf: https://linux.die.net/man/1/stdbuf
How about we use it for execution?    @mattgodbolt

**Comment by mattgodbolt** [comment]
I don't know if that would help here but we can give it a try and then do whatever syscalls it's doing rather than having to add stdbuf into the sandbox. Though that's possible too :)

**Comment by mattgodbolt** [comment]
Or...no it's more than that; https://hmarr.com/blog/how-stdbuf-works/ 

Which means it probably won't work for C++ streams.

**Comment by OfekShilon** [comment]
> Or...no it's more than that; https://hmarr.com/blog/how-stdbuf-works/
Great post! Thanks. It mentions in the end another option I wasn't aware of: https://linux.die.net/man/1/unbuffer

> Which means it probably won't work for C++ streams.

Why not? I believe the standard streams are the same for every process using glibc. Doesn't matter which API is used to change them.

1 more thought: unlike `stdbof`/`unbuffer` authors, _we have control over the source_ - so maybe we don't have to resort to LD_PRELOAD or pty heavy machinery. We can add to the source if execution is enabled something like -
```
#include <stdio.h>
__attribute__((constructor))
static void disable_io_line_buffering() { setvbuf(stdout, NULL, _IONBF, 0); }
```
This dows solve the issue (https://compiler-explorer.com/z/oEa8q5GG4) but feels a bit off. Eg, I wouldn't want this function to pollute asm/cfg views

**Comment by OfekShilon** [comment]
Yeah all in all my vote goes to `stdbuf`. @mattgodbolt can you please add it to the sandbox, so we could experiment with it?

**Comment by mattgodbolt** [comment]
Before we add it; can you check it actually works locally first? In sceptical that it works for anything other than libc's output

**Comment by OfekShilon** [comment]
You're right, I should do that

**Comment by OfekShilon** [comment]
@mattgodbolt I believe it works:

![Image](https://github.com/user-attachments/assets/1d609854-b077-4518-916d-f60547b6db8e)

Making `stdbuf` an `executionWrapper` as I did is one reasonable way to go about it, but we can discuss further. Could you please add stdbuf to nsjail when you have time?

**Comment by OfekShilon** [comment]
Actually, I better first check if it interacts properly with @jeremy-rifkin 's libSegFault

**Comment by OfekShilon** [comment]
Ok, libSegFault somehow works with stdbuf too:

![Image](https://github.com/user-attachments/assets/e27115c8-6022-4494-bc70-1a92870275bb)

**Comment by mattgodbolt** [comment]
sweeet!!! I'll see what it'd take to get stdbuf into the sandbox :tada:

**Comment by partouf** [comment]
But I'm assuming we're not adding executionWrapper literally everywhere, what's the next step exactly?

**Comment by OfekShilon** [comment]
> But I'm assuming we're not adding executionWrapper literally everywhere, what's the next step exactly?
> (I would have assumed we change https://github.com/compiler-explorer/nsjail to internally use stdbuf?)

I wasn't even aware we use our own nsjail fork..
So the alternatives I can see for using stdbuf:
1. Customize nsjail,
2. Use the existing executionWrapper config hook,
3. Hard-code stdbuf somewhere in `BaseCompiler.runExecutable` or below (if we go there, my vote goes to hard-code as low as possible, say in `executeDirect`)
4. Use a new config key defaulting to 'on', and test for it somewhere in `BaseCompiler.runExecutable` or below.

Some thoughts on tradeoffs:
- I don't think I'm in favor of customizing nsjail, as we'd want users of local instances to benefit from this too - AND be able to easily run the same setup locally ourselves while developing.
- AFAICT executionWrapper is used only in dart, and I don't know for what.  Adding it everywhere is indeed a nuisance, perhaps we can hard-code a default [when reading it](https://github.com/compiler-explorer/compiler-explorer/blob/a5aad3c48c6e1cb9655d7ed068c4a3e464e6512d/lib/compiler-finder.ts#L242-L243)
- I'm not aware of execution scenarios where `stdbuf` cannot/shouldn't be used but it's reasonable to assume such scenarios would come up, so it would be easier to prepare for them by making it configurable and not hard coded.
- I suspect it would be easiest to agree on using a new config key, something like `compiler.***.execUnbufferedStdout`.

WDYT?

**Comment by mattgodbolt** [comment]
I don't understand why we'd modify nsjail? It doesn't seem like we'd need to: nsjail doesn't care what it runs, and we're the ones telling it what to run...so we can much easier pass `stdbuf -o 0 -- <rest of commandline` or similar.

I'd be happy defaulting something in the execution properties or even making a new nsjailWithStdbuf execution type if that works.

It doesn't seem to be something that's compiler-specific (to me); it's an execution wrapper type thing (I get the dart annoyance).

So I guess my vote is 3 or 4.

**Comment by partouf** [comment]
> I don't understand why we'd modify nsjail? It doesn't seem like we'd need to: nsjail doesn't care what it runs, and we're the ones telling it what to run...so we can much easier pass `stdbuf -o 0 -- <rest of commandline` or similar.

Because it's already hard enough to track all the kinds of wrappers that stack upon each other as it is in CE. Would be nice if something would actually be more transparent and work without yet another wrapper.

But my main complaint was about adding it a million properties for this. Please, let's not.

**Comment by mattgodbolt** [comment]
100% agreed on a million properties. I see it as something to go in `exec.*.properties` globally myself

**Comment by mattgodbolt** [comment]
Fix confirmed to be working in staging! thank you!

**Comment by mattgodbolt** [comment]
Oh.
no
scratch that :( sorry - https://godbolt.org/staging/z/GMe1sfEGo still showing 👍🏻 

```
ASM generation compiler returned: 0
Execution build compiler returned: 0
Program returned: 139
aaa
Program terminated with signal: SIGSEGV
```

**Comment by narpfel** [comment]
Is `unbufferStdoutExe` missing in [`execution.amazon.properties`](https://github.com/compiler-explorer/compiler-explorer/blob/d3ce34dca67c64dfd5fd05a0ca9274ac2e93d94e/etc/config/execution.amazon.properties)?

**Comment by mattgodbolt** [comment]
> Is `unbufferStdoutExe` missing in [`execution.amazon.properties`](https://github.com/compiler-explorer/compiler-explorer/blob/d3ce34dca67c64dfd5fd05a0ca9274ac2e93d94e/etc/config/execution.amazon.properties)?

No, in this case it's inherited from the defaults. The fix is now being PRd: #7404  :) (thanks for taking a look!)

**Comment by mattgodbolt** [comment]
Ok! fixed for real now :)

**Comment by mattgodbolt** [comment]
Fix is now live!

**Comment by OfekShilon** [comment]
I forget what went wrong with stdbuf that caused us to revert that fix - but it was reverted and I'm reopening the issue.

Another [suggestion above](https://github.com/compiler-explorer/compiler-explorer/issues/7239#issuecomment-2585392438) was to add this to the source when execution is enabled:
```c
#include <stdio.h>
__attribute__((constructor))
static void disable_io_line_buffering() { setvbuf(stdout, NULL, _IONBF, 0); }
```

It might slightly pollute some views (asm, cfg) but on the tradeoff with losing output - I prefer that. WDYT?     @mattgodbolt

**Comment by partouf** [comment]
> I forget what went wrong with stdbuf that caused us to revert that fix - but it was reverted and I'm reopening the issue.

https://github.com/compiler-explorer/compiler-explorer/pull/7482/files
https://github.com/compiler-explorer/compiler-explorer/issues/7427

**Comment by mattgodbolt** [comment]
I still prefer us setting up a fake tty. Just needs a bit of a tweak in our sandbox, and a build of the fake tty binary.
